tfx_bsl/beam/run_inference.py

```diff
@@ -221,17 +223,129 @@ def update(self, elements: List[Union[tf.train.Example,
       self._inference_request_batch_byte_size.update(
           sum(element.ByteSize() for element in elements))
 
+  def __init__(self):
+    super(_BaseDoFn, self).__init__()
+    self._clock = None
+    self._metrics_collector = self._MetricsCollector()
+
+  def setup(self):
+    self._clock = _ClockFactory.make_clock()
+
+  def finish_bundle(self):
+    self._metrics_collector.update_metrics_with_cache()
+
+  def process(self, elements: list) -> Any:
+    batch_start_time = self._clock.get_current_time_in_microseconds()
+    result = self.run_inference(elements)
+    self._metrics_collector.update(
+        elements,
+        self._clock.get_current_time_in_microseconds() - batch_start_time)
+    return result
+
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> list:
+    raise NotImplementedError
+
+
+@beam.typehints.with_input_types(List[Union[tf.train.Example,
+                                            tf.train.SequenceExample]])
+class _RemotePredictDoFn(_BaseDoFn):
+  """A DoFn that performs predictions from a cloud-hosted model.
+
+  Supports both batch and streaming processing modes.
+  NOTE: Does not work on DirectRunner for streaming jobs [BEAM-7885].
+
+  In order to request predictions, you must deploy your trained model to AI
+  Platform Prediction. See
+  https://cloud.google.com/ai-platform/prediction/docs/online-predict for more
+  details.
+  """
+  def __init__(self, inference_endpoint: model_spec_pb2.InferenceEndpoint):
+    super(_RemotePredictDoFn, self).__init__()
+    self._api_client = None
+    project_id = inference_endpoint.model_endpoint_spec.project_id
+    if not project_id:
+      raise ValueError('A non-empty project id must be provided.')
+
+    model_name = inference_endpoint.model_endpoint_spec.model_name
+    if not model_name:
+      raise ValueError('A non-empty model name must be provided.')
+
+    version_name = inference_endpoint.model_endpoint_spec.version_name
+    name_spec = 'projects/{}/models/{}'
+    # If version is not specified, the default version for a model is used.
+    if version_name:
+      name_spec += '/versions/{}'
+    self.full_model_name = name_spec.format(project_id, model_name,
+                                            version_name)
+
+  def setup(self):
+    super(_RemotePredictDoFn, self).setup()
+    self._api_client = self._get_api_client()
+
+  @staticmethod
+  def _get_api_client() -> discovery.Resource:
+    return discovery.build('ml', 'v1')
+
+  @staticmethod
+  def _execute_request(request: http.HttpRequest) -> dict:
+    return request.execute()
+
+  def _make_request(self, model_name: str, body: dict) -> http.HttpRequest:
+    return self._api_client.projects().predict(name=model_name, body=body)
+
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> list:
```

Comment:
What is the normal return value from this function?

Reply:
It returns a list of JSON objects which could be of any complexity: it depends on model. For instance, according to docs, if model has more than one output tensor, each element in the list is a dictionary containing name/value pairs.

I'll try to convert the response to PredictLog objects.